src/pool/srv_pool.c

```diff
@@ -447,6 +447,87 @@ pool_prop_write(struct rdb_tx *tx, const rdb_path_t *kvs, daos_prop_t *prop)
 	return rc;
 }
 
+static int
+pool_buf_fill_extend(struct pool_map *map, struct pool_buf *map_buf,
+		int nnodes, int map_version, int ndomains,
+		const int32_t *domains, uuid_t target_uuids[],
+		const d_rank_list_t *target_addrs, uuid_t *uuids, bool *updated)
+{
+	struct pool_component	map_comp;
+	uint32_t		num_comps;
+	struct pool_domain	*found_dom;
+	int i, rc;
+
+	num_comps = pool_map_find_domain(map, PO_COMP_TP_RACK, PO_COMP_ID_ALL,
+					 NULL);
+	/* fill domains */
+	for (i = 0; i < ndomains; i++) {
+		map_comp.co_type = PO_COMP_TP_RACK;	/* TODO */
+		map_comp.co_status = PO_COMP_ST_NEW;
+		map_comp.co_index = i;
+		map_comp.co_id = i + num_comps;
+		map_comp.co_rank = 0;
+		map_comp.co_ver = map_version;
+		map_comp.co_fseq = 1;
+		map_comp.co_nr = domains[i];
+
+		rc = pool_buf_attach(map_buf, &map_comp, 1 /* comp_nr */);
+		if (rc != 0)
+			return rc;
+	}
+
+	num_comps = pool_map_find_domain(map, PO_COMP_TP_NODE, PO_COMP_ID_ALL,
+					 NULL);
+	/* fill nodes */
+	for (i = 0; i < nnodes; i++) {
+		uuid_t *p = bsearch(target_uuids[i], uuids, nnodes,
+				    sizeof(uuid_t), uuid_compare_cb);
+
+		found_dom = pool_map_find_node_by_rank(map, domains[i]);
+		if (found_dom)
+			continue;
+
+		*updated = true;
+		map_comp.co_type = PO_COMP_TP_NODE;
+		map_comp.co_status = PO_COMP_ST_NEW;
+		map_comp.co_index = i;
+		map_comp.co_id = (p - uuids) + num_comps;
+		map_comp.co_rank = target_addrs->rl_ranks[i];
+		map_comp.co_ver = map_version;
+		map_comp.co_fseq = 1;
+		map_comp.co_nr = dss_tgt_nr;
+
+		rc = pool_buf_attach(map_buf, &map_comp, 1 /* comp_nr */);
+		if (rc != 0)
+			return rc;
+	}
+
+	if (!updated)
+		return 0;
+
+	num_comps = pool_map_find_target(map, PO_COMP_ID_ALL, NULL);
+	/* fill targets */
+	for (i = 0; i < nnodes; i++) {
+		int j;
+
+		for (j = 0; j < dss_tgt_nr; j++) {
+			map_comp.co_type = PO_COMP_TP_TARGET;
+			map_comp.co_status = PO_COMP_ST_NEW;
+			map_comp.co_index = j;
+			map_comp.co_id = (i * dss_tgt_nr + j) + num_comps;
+			map_comp.co_rank = target_addrs->rl_ranks[i];
+			map_comp.co_ver = map_version;
+			map_comp.co_fseq = 1;
+			map_comp.co_nr = 1;
+
+			rc = pool_buf_attach(map_buf, &map_comp, 1);
+			if (rc != 0)
+				return rc;
+		}
+	}
```

Comment:
These are very similar as one part of init_pool_metadata(). Could you please generate a common function?

Reply:
Refactored into a single function. See gen_pool(...)